src/main/java/org/betonquest/betonquest/objectives/CommandObjective.java

```diff
@@ -0,0 +1,138 @@
+        commandVariables = loadVariables(command);
+        ignoreCase = instruction.hasArgument("ignoreCase");
+        exact = instruction.hasArgument("exact");
+        cancel = instruction.hasArgument("cancel");
+        failEvents = instruction.getList(instruction.getOptional("failEvents"), instruction::getEvent).toArray(new EventID[0]);
+    }
+
+    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
+    public void onChat(final AsyncPlayerChatEvent event) {
+        if (event.isCancelled() && event.getMessage().startsWith("/")) {
+            // A different plugin intercepted this chat "command" already, handle unregistered command
+            checkForRequiredCommand(event.getPlayer(), event.getMessage(), event);
+        }
+    }
+
+    @EventHandler(priority = EventPriority.LOWEST)
+    public void onCommand(final PlayerCommandPreprocessEvent event) {
+        checkForRequiredCommand(event.getPlayer(), event.getMessage(), event);
+    }
+
+    @Override
+    public void start() {
+        Bukkit.getPluginManager().registerEvents(this, BetonQuest.getInstance());
+    }
+
+    @Override
+    public void stop() {
+        HandlerList.unregisterAll(this);
+    }
+
+    @Override
+    public String getDefaultDataInstruction() {
+        return "";
+    }
+
+    @Override
+    public String getProperty(final String name, final String playerID) {
+        return "";
+    }
+
+    // TODO: This is mainly a functional copy-paste from NotifyEvent. Variable handling should eventually be standardized.
+    private List<String> loadVariables(final String message) throws InstructionParseException {
+        final List<String> variables = new ArrayList<>();
+        for (final String variable : BetonQuest.resolveVariables(message)) {
+            try {
+                BetonQuest.createVariable(instruction.getPackage(), variable);
+            } catch (final InstructionParseException exception) {
+                throw new InstructionParseException("Could not create '" + variable + "' variable: "
+                        + exception.getMessage(), exception);
+            }
+            if (!variables.contains(variable)) {
+                variables.add(variable);
+            }
+        }
+        return variables;
+    }
+
+    @SuppressWarnings("PMD.AvoidDeeplyNestedIfStmts")
+    private void checkForRequiredCommand(final Player player, final String commandExecuted, final Cancellable event) {
+        final String playerID = PlayerConverter.getID(player);
+        if (containsPlayer(playerID) && checkConditions(playerID)) {
+            final String replaceCommand = getCommandWithVariablesReplaced(playerID);
+            if (foundMatch(commandExecuted, replaceCommand)) {
+                if (cancel) {
+                    event.setCancelled(true);
+                }
+                Bukkit.getScheduler().runTask(BetonQuest.getInstance(), () -> completeObjective(playerID));
```

Comment:
```suggestion
                completeObjective(playerID);
```
call it directly, the sync or async execution is already handled by the events itself